### ISSUE TITLE
Migrate from marked to markdown-it

### ISSUE THREAD

**Issue by leipert** [Galadirith/markdown-preview-plus#47]
An alternative to `marked` could be [`markdown-it`](https://github.com/markdown-it/markdown-it).

The main advantages of markdown-it as I see them:

* Follows [CommonMark spec](https://www.npmjs.com/browse/keyword/markdown-it-plugin)
* Has the possibility to write plugins for it, so that we would not need dirty hacks. ([a lot of them exist already](https://www.npmjs.com/browse/keyword/markdown-it-plugin))

Furthermore, we actually could get pandoc compatibility without having to use pandoc!

**Comment by leipert** [comment]
We could also make configurable, which plugins to use in the settings.
We actually could install all plugins listed below and just lazy-load them if needed.

The plugins I found most useful for incorporation.

* [Code highlighting](https://www.npmjs.com/package/markdown-it-highlightjs) with highlightjs
* [Math Rendering](https://github.com/runarberg/markdown-it-math), does not use mathjax, but  could be possibly adjusted to use it 
* [Definitions Lists](https://www.npmjs.com/package/markdown-it-deflist) like in [pandoc](http://pandoc.org/README.html#definition-lists)
* [Footnotes](https://www.npmjs.com/package/markdown-it-footnote) like in [pandoc](http://pandoc.org/README.html#footnotes)
* [sub](https://www.npmjs.com/package/markdown-it-sub) and [sup](https://www.npmjs.com/package/markdown-it-sup) like in [pandoc](http://pandoc.org/README.html#superscripts-and-subscripts)
* Maybe automatic [table of contents](markdown-it-toc)? 

There are much more, also some generic ones that allow general regex replacements.

At the moment I don't miss any major pandoc features, but if we switch to `markdown-it`, I would take care of creating plugins that allow full pandoc compatibility.

Well. That's it for now, I will go back to studying as I have an exam in three hours

**Comment by Galadirith** [comment]
@leipert Thanks so much for compiling this info. After a first look I would be very happy to migrate to [`markdown-it`](https://github.com/markdown-it/markdown-it), definitely commonMark compliance and extendibility through plugins makes it super attractive. 

A couple of thoughts about the plugins:

- [`markdown-it-highlightjs`](https://www.npmjs.com/package/markdown-it-highlightjs) might not be as useful. [`markdown-preview`](https://github.com/atom/markdown-preview) has for a while been replacing `<pre>` blocks with `<atom-text-editor>`. The reason is that highlighting then uses the syntax theme configured in Atom to ensure a unified looked. I chose not to include that change when it was committed upstream because I wasn't sure how it would interfere with the diff algorithm, but having looked at the code I think its fine, so that can probably come out in the next release with a few tweaks. `markdown-preview` handles this as a post process to the markdown parsing so is currently independent of the markdown parser, but it might possibly we worth adding it as a plugin in time.

- [`markdown-it-math`](https://github.com/runarberg/markdown-it-math) looks like a good place to start. However I certainly would want to make sure we've got a plugin that can properly support LaTeX maths expressions to MathJax `<script>` blocks (which I know my current hack of `marked` doesn't, #20, #17, #7 to name a few). But I definitely much much prefer the idea of maintaining a plugin rather than a hacked fork. 

I will hopefully have time over the weekend to familiarize myself with `markdown-it`'s API and I would be very happy to have a go at putting together a MathJax plugin.

I really hope everything went well for you in your exam, thanks again @leipert some really exciting stuff here :D

**Comment by leipert** [comment]
See 20974f6 for my first try with `markdown-it`. It is quite easy to get it to work with Mathjax.
The diff-algorithm does not to seem working well with it though.

For a roadmap I would really like the following steps:

1. Define a lot more tests, that way we ensure, we will not break anything that is working.
2. Define tests for all issues that already arose already, like #7, #17, and #20 (they work with markdown-it by the way)
3. Make a list of all markdown/pandoc features we want to have
4. It would be really cool, if the package is configurable for the user. You could enable each feature with a `+` and disable with a `-`, like `+tex_math_dollars-tex_math_single_backslash+footnotes-definition_lists`, would enable Math syntax with dollar and footnotes. Math with brackets and definition lists would be disabled.

**Comment by Galadirith** [comment]
That's incredible :D I'm really sorry I can't respond to the roadmap right now (other than it looks pretty solid, but I'll give you a better reply later) but I've push a fix dfc9b17 to make `markdown-it` compatible with the diff-algorithm.

**Comment by Galadirith** [comment]
@leipert I very sorry for the lack of communication over the last couple of weeks.

Everything that you said looks very sensible. Would you be happy to handle this migration? While you handle this I will work on implementing #9 which I've already begun and also write up specs that I've rather foolishly ignored for too long; `mathjax-helper`, `update-preview`. I will also handle writing specs for #7, #17, and #20 and more. Would you be happy this that?

There was one thought I had. `markdown-it-math` distinguishes inline maths as embedded in paragraphs, and displayed maths as external to paragraphs (separated by blank lines from other block level elements). As discussed in #2 LaTeX also allows displayed maths to be embedded in paragraphs, so it would also need to support `$$` and `\[` as in-line delimiters. I do think its important to support this however `markdown-it-math` while supporting [multiple types of delimiter](https://github.com/runarberg/markdown-it-math/blob/master/index.js#L218-L219) only creates a single inline maths rendering rule [math_inline](https://github.com/runarberg/markdown-it-math/blob/master/index.js#L220) no matter how many times you `markdownIt.use` the `markdown-it-math` package.

**Comment by leipert** [comment]
@Galadirith I suppose it is alright for now, that `$$` and `\[` are not supported as inline delimiters for the following reasons and I will try to defend `markdown-it-maths` behavior (for now):

1. They are block delimiters.
2. It is generally a good style to separate markdown leaf blocks with blank lines before and after for readability.
3. Math is not part of the commonmark spec, so why not enforce good standards regarding the difference between `$` and `$$`

**Comment by leipert** [comment]
Furthermore this regression is okay, as we are still aligning with the semver model -> update to 2.0.0 and there is this breaking change which enforces a nice style :+1:

**Comment by Galadirith** [comment]
@leipert Thanks so much for your thoughtful comments. I do understand your points and your semver comment does make a lot of sense in this context.

> 1. They are block delimiters.
> 2. It is generally a good style to separate markdown leaf blocks with blank lines before and after for readability.
> 3. Math is not part of the commonmark spec, so why not enforce good standards regarding the difference between `$` and `$$`

I completely agree with on point (2) and partially on point (3).

**TL;DR** (Markdown block level elements) != (Displayed LaTeX equations)

My argument would be that actually when `$$` is embedded in a paragraph I really do mean it to be an inline element and not a block level element. LaTeX does actually distinguish between displayed maths isolated by empty lines from other content and displayed maths not isolated, so `doc1.tex`

```
% doc1.tex
askajsdlak asdas asdasd $$ \int^3_4$$ asdasd asdasd
```

is actually rendered differently than `doc2.tex`

```
% doc2.tex
asdasd asd asd 

$$
\int^3_4
$$

asdasdas asdas 
```

and the rendering is controlled by lengths such as `\parskip`.

I have whipped up a [fork of `markdown-it-math`](https://github.com/Galadirith/markdown-it-math/tree/feature/inline-display-math) that handles this, and doesn't break any of its original tests, but I haven't yet add tests for the new feature. `markdown-it` integration has been your work, so your call if you want to leave this out for now and I'll work on making sure this is stable for a 2.x.x release.

That all said, I actually never use `$$` embedded in paragraphs, I do think it should be treated solely as a block level element. However LaTeX has been around for a very long time, longer than markdown and my feeling is that people do what this feature, but perhaps leaving such a feature out for `2.0.0` will be a good gauge for its desirability :D

Thanks @leipert :D

**Comment by leipert** [comment]
I tried your fork, it dies not fix it unfortunately, as you can see with some tests failing: b9ae65acd926ad797ff2f1781469914ce4d591a5

Do you want to try to fix those mistakes in markdown-it-math or do we want to live with those things not working now?

**Comment by Galadirith** [comment]
I happy to not have that feature for now, as you pointed out this regression is perfectly in-line with sumver, and I wouldn't want to spend more than an evening making sure everything is working and stable. Lets run without it for `2.0.0`.
